src/libsyntax/ext/deriving/generic/mod.rs

```diff
@@ -549,10 +550,20 @@ impl<'a> TraitDef<'a> {
                 .map(|ty_param| ty_param.ident.name)
                 .collect();
 
+            let mut processed_field_types = HashSet::new();
             for field_ty in field_tys {
                 let tys = find_type_parameters(&*field_ty, &ty_param_names);
 
                 for ty in tys {
+                    // if we have already handled this type, skip it
+                    if let ast::TyPath(_, ref p) = ty.node {
+                        if p.segments.len() == 1
+                            && ty_param_names.contains(&p.segments[0].identifier.name)
```

Comment:
I don't think so, unless I'm misunderstanding your comment. `ty_param_names` are `ast::Name`s, while the elements of `processed_field_types` are `Vec<ast::PathSegment>`. It's not clear to me that we can somehow convert one into the other for comparison.